Mathlib/MeasureTheory/Function/L1Space.lean

```diff
@@ -1443,6 +1481,62 @@ theorem toL1_smul' (f : α → β) (hf : Integrable f μ) (k : 𝕜) :
 
 end Integrable
 
+section restrict
+
+variable {E : Type*} [NormedAddCommGroup E] {f : α → E}
+
+lemma HasFiniteIntegral.restrict (h : HasFiniteIntegral f μ) {s : Set α} :
+    HasFiniteIntegral f (μ.restrict s) := by
+  refine lt_of_le_of_lt ?_ h
+  convert lintegral_mono_set (μ := μ) (s := s) (t := univ) (f := fun x ↦ ↑‖f x‖₊) (subset_univ s)
+  exact Measure.restrict_univ.symm
+
+lemma Integrable.restrict (f_intble : Integrable f μ) {s : Set α} :
+    Integrable f (μ.restrict s) :=
+  ⟨f_intble.aestronglyMeasurable.restrict, f_intble.hasFiniteIntegral.restrict⟩
+
+lemma ae_restrict_eq_zero_iff_ae_eq_zero_of_mem [MeasurableSpace E] [MeasurableSingletonClass E]
+    {s : Set α} (f_mble : NullMeasurable f (μ.restrict s)) :
+    f =ᵐ[Measure.restrict μ s] 0 ↔ ∀ᵐ x ∂μ, x ∈ s → f x = 0 := by
+  simp only [Measure.ae, MeasurableSet.compl_iff, EventuallyEq, Filter.Eventually,
+             Pi.zero_apply, Filter.mem_mk, mem_setOf_eq]
+  rw [Measure.restrict_apply₀]
+  · constructor <;> intro h <;> rw [← h] <;> congr <;> ext x <;> aesop
+  · apply NullMeasurableSet.compl
+    convert f_mble (MeasurableSet.singleton 0)
+
+lemma ae_restrict_eq_zero_iff_ae_eq_zero_of_mem' {s : Set α} (s_mble : MeasurableSet s) :
+    f =ᵐ[Measure.restrict μ s] 0 ↔ ∀ᵐ x ∂μ, x ∈ s → f x = 0 := by
+  simp only [Measure.ae, MeasurableSet.compl_iff, EventuallyEq, Filter.Eventually,
+             Pi.zero_apply, Filter.mem_mk, mem_setOf_eq]
+  rw [Measure.restrict_apply_eq_zero']
+  · constructor <;> intro h <;> rw [← h] <;> congr <;> ext x <;> aesop
+  · exact s_mble
+
+lemma NullMeasurable.measure_preimage_eq_measure_restrict_preimage_of_ae_compl_eq_const
+    {β : Type*} [MeasurableSpace β] {b : β} {f : α → β} {s : Set α}
+    (f_mble : NullMeasurable f (μ.restrict s)) (hs : f =ᵐ[Measure.restrict μ sᶜ] (fun _ ↦ b))
+    {t : Set β} (t_mble : MeasurableSet t) (ht : b ∉ t) :
+    μ (f ⁻¹' t) = μ.restrict s (f ⁻¹' t) := by
+  rw [Measure.restrict_apply₀ (f_mble t_mble)]
+  simp only [EventuallyEq, Filter.Eventually, Pi.zero_apply, Measure.ae,
+             MeasurableSet.compl_iff, Filter.mem_mk, mem_setOf_eq] at hs
+  rw [Measure.restrict_apply₀] at hs
+  · apply le_antisymm _ (measure_mono (inter_subset_left _ _))
```

Comment:
I am not sure these 3 lemmas belong here, as they don't involve integrable functions. Move them to a more appropriate place?

Reply:
Indeed, thanks. I moved to `MeasureSpace.lean`, which is where `Measure.restrict` is defined.

Reply:
An update: I had to move two of these three to `SetIntegral.lean`, because `MeasureSpace.lean` was not imported there and they were used. But I suppose that file is still a better place than `L1Space.lean`.